src/nnet3/nnet-simple-component.h

```diff
@@ -873,6 +875,68 @@ class PerElementScaleComponent: public UpdatableComponent {
 };
 
 
+// PerElementOffsetComponent offsets each dimension of its input with a separate
+// trainable bias; it's like an affine component with fixed weight matrix which is always equal to I.
+class PerElementOffsetComponent: public UpdatableComponent {
+ public:
+  virtual int32 InputDim() const { return offsets_.Dim(); }
+  virtual int32 OutputDim() const { return offsets_.Dim(); }
+
+  virtual std::string Info() const;
+  virtual void InitFromConfig(ConfigLine *cfl);
+
+  PerElementOffsetComponent() { } // use Init to really initialize.
+  virtual std::string Type() const { return "PerElementOffsetComponent"; }
+  virtual int32 Properties() const {
+    return kSimpleComponent|kUpdatableComponent|
+           kBackpropInPlace|kPropagateInPlace;
```

Comment:
Please remove the kPropagateInPlace flag, as your propagation code would not work correctly if &in == out.
I just realized that we are not testing the kPropagateInPlace flag; I am adding testing code for this.